### ISSUE TITLE
Discussion: Routable IP addresses

### ISSUE THREAD

**Issue by robertvazan** [appc/spec#294]
AppContainer repeats the mistake of Docker of restricting containers to NAT-ed ports. This prevents services from running on standard ports and prevents configuration of clients via DNS.

At the very least, it should be possible to map container ports to specific IP address on the host. But that's not enough, because it makes it hard to deploy the apps. One has to first configure something on the host via ssh. Not good. There must be some way to auto-configure it.

If the network has DHCP, perhaps ACE could just ask for new IP address, sending generated/virtual MAC address to the DHCP server.

IPv6 has some auto-configuration options that would allow assigning world-routable IPv6 address to every container. AppContainer is exactly the kind of system that could make good use of IPv6.

**Comment by jonboulle** [comment]
> AppContainer repeats the mistake of Docker of restricting containers to NAT-ed ports. 

I don't understand how you reached this conclusion, can you elaborate? The only thing we stipulate in the spec right now is that each Pod must have its own layer 3 network interface, specifically so that _can_ listen on standard ports. Perhaps we should tweak the text to clarify this?

I think most of your questions (like mapping ports from the host) are more specific to particular implementations than what we should discuss in the spec. Can you please take a look at rkt's networking implementation and see if this clarifies/addresses your concerns?
https://github.com/coreos/rkt/blob/master/Documentation/networking.md

**Comment by robertvazan** [comment]
Indeed rkt implements routable IP addresses at least in IPv4. What I am looking for is a way to specify static IP address in POD and a way to request dynamic IP address from ACE via some declaration in the POD.

**Comment by eyakubovich** [comment]
You're correct in that rkt currently implements only IPv4. We do want to support IPv6 but since v6 is still far from ubiquitous, we need to get v4 working first. In the data center where RFC 1918 addresses are the norm, there's not much contention on the address space.

AppC spec is perfectly fine with each pod getting a world routable address but it does not require it. Some implementations and environments maybe able to fulfill that wish. rkt's networking is plugin-based for flexibility because every environment is different (here's current proposal that's being implemented: https://docs.google.com/a/coreos.com/document/d/1PUeV68q9muEmkHmRuW10HQ6cHgd4819_67pIxDRVNlM)

If you would like to use macvlan (not all environments support it), here's the plugin for that: https://github.com/coreos/rkt/blob/master/networking/net/macvlan/macvlan.go You can pair it with DHCP for address assignment (plugin under development: https://github.com/eyakubovich/rkt/tree/ipam-dhcp/networking/ipam/dhcp). An alternative would be to use overlay networking via linux-bridge + flannel or weave. Not all plugins are done but the goal is to provide multiple options for making each pod having a routable IP a reality.

**Comment by robertvazan** [comment]
Hmmm. Let me think about it. If ACE assigns the POD an IP address that is already routable, then the POD doesn't need port mapping at all, am I right? So I just exclude the port mapping from the POD and let the app bind on the routable IP address, right?

If it can be configured this way, then indeed it is unnecessary to add extra configuration options in the POD. I can just preconfigure all nodes to hand routable IPs to containers and forget the whole port mapping altogether.

I misread the spec to mean that the port mapping is necessary to make anything visible on the network. I am sorry, but at least we have public FAQ about it.

**Comment by jonboulle** [comment]
> Hmmm. Let me think about it. If ACE assigns the POD an IP address that is already routable, then the POD doesn't need port mapping at all, am I right? So I just exclude the port mapping from the POD and let the app bind on the routable IP address, right?

Yes, exactly. The ideal scenario is that everything is aware of the network and a bare minimum of port mapping is necessary. @eyakubovich touched on this in the blog for rkt 0.5.3 (our first release with port mapping): 

>> Whenever possible, it is more convenient to use a SDN solution like flannel to assign routable IPs to rkt pods. However, when such an option is not available, or for "edge" apps that require straddling both SDN and external networks (such as a load balancer), port forwarding can be used to expose select ports to the pod.


.
  


> I misread the spec to mean that the port mapping is necessary to make anything visible on the network. I am sorry, but at least we have public FAQ about it.

Do you think you could put up a PR to clarify? Thanks!

**Comment by krobertson** [comment]
It might be good to clarify port mapping is to map in host ports to the pod, potentially.

We're looking at doing a L2 network between pods, so a group of pods can communicate easily, but to not have that network routable so we still control outbound egress and ingress separately. Basically allowing clustered services, but still controlling who can get to the service.
